File: C/Horspool's Algorithm/horspool.c
#include<stdio.h>
#include<string.h>
#define MAX 500
int t[MAX];
void shifttable(char p[])
{
	int i,j,m;
	m=strlen(p);
	for(i=0;i<MAX;i++)
	{
		t[i]=m;
	}
	for(j=0;j<m-1;j++)
	{
		t[p[j]]=m-1-j;
	}
}
int horspool(char src[],char p[])
{
	int i,j,k,m,n;
	n=strlen(src);
	m=strlen(p);
	i=m-1;
	while(i<n)
	{
		k=0;
		while((k<m)&& p[m-1-k]==src[i-k])
		{
			k++;
		}
		if(k==m)
		{
			return(i-m+1);
		}
		else
		{
			i+=t[src[i]];
		}
	}
	return -1;
}
void main()
{
	char src[100],p[100];
	int pos;
	printf("enter string\n");
	gets(src);
	printf("enter the pattern\n");
	gets(p);
	shifttable(p);
	pos=horspool(src,p);
	if(pos>=0)
	{
		printf("the pattern begins from the position %d\n",pos);
	}
	else
	{
		printf("the pattern doesnt exist\n");
	}
}
	

File: Java/Horspool's Algorithm/Horspool.java
import java.util.*;
public class Horspool  {
  
   public static int SIZE=500;    
   public static int table[]=new int[SIZE];
  
public void shifttable(String pattern) {
 
 int i,j,m;
 char p[] = pattern.toCharArray();
 m=pattern.length();
 
 for (i=0;i<SIZE;i++)
    table[i]=m;
 for (j=0;j<m;j++)
    table[p[j]]=m-1-j;
}
public int horspool(String source,String pattern)
  {
      int i,k,pos,m;
      char s[] = source.toCharArray();
      char p[] = pattern.toCharArray();
      m=pattern.length();
     
      for(i=m-1;i<source.length();)
        {
           k=0;
            while((k<m) && (p[m-1-k] == s[i-k]))
              k++;
           if(k==m) 
     {   pos=i-m+2;
               return pos;
     } 
           else
               i+=table[s[i]];
        }
        return -1;
  } 
     public static void main(String []args){
        int pos; 
  String source=args[0];
        String pattern = args[1];
  
        Horspool  h = new Horspool ();       
        h.shifttable(pattern);
        pos = h.horspool(source,pattern);

        if(pos == -1)
            System.out.println("PATTERN NOT FOUND");
        else
            System.out.println("PATTERN FOUND FROM POSITION: \t"+pos+"\n");
     }
}